### PR TITLE
IDCOM-2126 Add an optional failsafe key to the checkPass middleware

### PR DESCRIPTION
**Pull request by dankelleher**
When creating a middleware account for the checkPass middleware, a failsafe key can be added which bypasses the pass check.
This _cannot_ be changed after the middleware is created, and must be a key on the DID (otherwise the signing of the overall cryptid tx will fail.

### CODE REVIEW

File: packages/tests/src/middleware/chaining.ts [hunk 1]
```diff
@@ -0,0 +1,249 @@
+// import {
```
Comment (rado0x54) [inline packages/tests/src/middleware/chaining.ts:1]: Should we commit this yet?

File: programs/middleware/check_pass/src/lib.rs [hunk 2]
```diff
@@ -169,6 +195,7 @@ impl<'info> ExecuteMiddleware<'info> {
             CheckPass::SEED_PREFIX,
             authority_key.as_ref(),
             gatekeeper_network_key.as_ref(),
+            ctx.accounts.middleware_account.failsafe.as_ref().map(|f| f.as_ref()).unwrap_or(&[0u8; 32]),
```
Comment (rado0x54) [inline programs/middleware/check_pass/src/lib.rs:198]: Don't completely understand this handling of a Option<Pubkey>. Why can you not use unwrap_or directly?

File: programs/middleware/check_pass/src/lib.rs [hunk 3]
```diff
@@ -105,15 +127,18 @@ pub mod check_pass {
 gatekeeper_network: Pubkey,
 /// The bump seed for the middleware signer
 bump: u8,
-/// Expire a gateway token after it has been used
-expire_on_use: bool
+/// Expire a gateway token after it has been used. Note, this can only be used
+/// with gatekeeper networks that have the ExpireFeature enabled.
+expire_on_use: bool,
+/// A key which, if passed as the authority, bypasses the middleware check
+failsafe: Option<Pubkey>
```
Comment (rado0x54) [inline programs/middleware/check_pass/src/lib.rs:134]: Do we need the Optional logic here? Could we not default it to [0u8; 32] during creating if not passed?

Comment (rado0x54) [reply programs/middleware/check_pass/src/lib.rs:134]: ah wait, ok, then it would never match, forget this comment.

**Review by rado0x54 — APPROVED** [review]

File: packages/tests/src/middleware/chaining.ts (hunk 1 above)
Comment (dankelleher) [reply packages/tests/src/middleware/chaining.ts:1]: fixed in the next pr - apologies I didn't realise I had it here.

File: programs/middleware/check_pass/src/lib.rs (hunk 2 above)
Comment (dankelleher) [reply programs/middleware/check_pass/src/lib.rs:198]: In summary, I need both as_refs.

- as_rev 1 turns the &Option<Pubkey> to an Option<&Pubkey>
- This allows the map to map the option and return (using as_ref again) the internal pubkey as an u8 slice (if I don't do that, map still works but you get moving issues)

an alternative would be to do unwrap_or(Pubkey(0u8; 30)) (or similar) but that would not pass as 0000... is not a valid pubkey. So you have to extract it into a byte slice first before doing the unwrap.

**Pull request merged**